---
output: html_document
params: 
  MAF:
    label: "Input file: "
    value: "CBIO.MAF"
    input: file
  author: "Stefan Kirov"
  title: "MAFMNV Checker run"
---

---
title: `r params$title`
author: `r params$author`
date: `r format(Sys.time(), '%d %B, %Y')`
---

```{r setup, echo=FALSE, warning=FALSE, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(shiny.maxRequestSize=5*1024^3)
options(rsconnect.max.bundle.size = 5*1024^3)
library(ggplot2)
library(reshape)
library(Biostrings)
library(kableExtra)

```

```{r ReadAndCheck, echo=FALSE}
columnscheck<-c("Symbol","Sample","Transcript","protein_position","HGVSp","t_ref_count","t_alt_count","Codons","Consequence") 

mafD<-read.table(params$MAF,sep="\t",head=T,quote = "",colClasses = "character")
print("Size of input set across all samples")
print(dim(mafD))

colindex<-lapply(columnscheck,function(x) grep(x,colnames(mafD),ignore.case = T))
names(colindex)<-columnscheck
colsfound<-all(unlist(lapply(colindex,function(x) length(x)>0)))

#If we cannot find one of the columns exit
if (!colsfound) {
   knitr::knit_exit("Error: One of the required columns is not parseable </body></html>")
}

#Extract actual names
symbolcol<-colnames(mafD)[colindex[["Symbol"]][1]]
samplecol<-colnames(mafD)[colindex[["Sample"]][1]]
hgvspcol<-colnames(mafD)[colindex[["HGVSp"]][1]]
transcriptcol<-colnames(mafD)[colindex[["Transcript"]][1]]
positioncol<-colnames(mafD)[colindex[["protein_position"]][1]]
trefcntcol<-colnames(mafD)[colindex[["t_ref_count"]][1]]
taltcntcol<-colnames(mafD)[colindex[["t_alt_count"]][1]]
consequencecol<-colnames(mafD)[colindex[["Consequence"]][1]]
codonscol<-colnames(mafD)[colindex[["Codons"]][1]]

```


```{r MNVfunctions, echo=FALSE}


combineMNVs<-function(rowsMNV) {
  #Assume 2 only for now
  #If these are the same substitutions they may be merged or otherwise ineligible
  if (rowsMNV[1,6+colindex[['HGVSp']][1]]==rowsMNV[2,6+colindex[['HGVSp']][1]]) {
    return(NULL)
  }
  rowData<-c(rowsMNV[1,c(1,2,3,4,5,6,6+colindex[['t_ref_count']][1],6+colindex[['t_alt_count']][1],6+colindex[['HGVSp']][1],6+colindex[['Codons']][1])],rowsMNV[2,c(6+colindex[['t_ref_count']][1],6+colindex[['t_alt_count']][1],6+colindex[['HGVSp']][1],6+colindex[['Codons']][1])])
  #rowData<-c(rowsMNV[1,c(1,2,3,4,5,6,40,41,45,50)],rowsMNV[2,c(40,41,45,50)])
  names(rowData)<-c("UID","Gene","Sample","Transcript","Position","Count","MutReads1","RefReads1","Mutation1","Codon1","MutReads2","RefReads2","Mutation2","Codon2")
  return(unlist(rowData))
}

getMNVs<-function(uidlist,allcodons.mnv) {
  mnv.data<-lapply(uidlist, function(x) combineMNVs(allcodons.mnv[allcodons.mnv$UID==x,]))
  return(Filter(Negate(is.null), mnv.data))
}

combineCodons<-function (codon) {
  icodons<-strsplit(codon[[10]],"/")
  ivec1<-strsplit(icodons[[1]][2],"")
  icodons<-strsplit(codon[[14]],"/")
  ivec2<-strsplit(icodons[[1]][2],"")
  
  finalvec<-vector()
  for (i in 1:3) {
    if  (ivec1[[1]][i]==toupper(ivec1[[1]][i])) {
      finalvec[i]<-ivec1[[1]][i]
    }
    else {
       finalvec[i]<-ivec2[[1]][i]
    }
  }
newcodon<-toupper(paste0(finalvec, collapse = ''))
return(c(newcodon,GENETIC_CODE[[newcodon]]))
}

determineConsequences<-function(conseq) {
  mut1v<-strsplit(conseq[[9]],"\\d+")
  mut2v<-strsplit(conseq[[13]],"\\d+")
  fmut<-conseq[[19]]
  mut1<-mut1v[[1]][2]
  mut2<-mut2v[[1]][2]
  orimut<-mut1v[[1]][1] 
  orimut<-gsub("p\\.","",orimut)
  
  if (mut1=="Ter") mut1="*"
  if (mut2=="Ter") mut2="*"
  if (mut1==fmut | mut1==fmut) {
   return("Same as one independent SNV")
  }
  if (fmut==orimut) {
    return("Reversal of mutation")
  }
  if (mut1=='*' | mut2=='*' | fmut == '*') {
    return("Change in stop codon")
  }
  return("Novel amino acid")
}

```

```{r analyzeMAF, echo=FALSE}

mafD<-mafD[mafD[,consequencecol]=="synonymous_variant"|mafD[,consequencecol]=="missense_variant"|mafD[,consequencecol]=="stop_gained"|mafD[,consequencecol]=="stop_lost",]
mafD<-mafD[grep("^IG[HKL]",mafD[,symbolcol],invert=T),]
mafD<-mafD[grep("^HLA",mafD[,symbolcol],invert=T),]
print("Size of input set across all samples after filtering")
print(dim(mafD))
aggregateform<-as.formula(paste(hgvspcol," ~ ",symbolcol,"+",samplecol,"+",transcriptcol,"+",positioncol))
allcodons.stats<-aggregate(aggregateform,data=mafD,length)
allcodons.stats<-allcodons.stats[allcodons.stats[,5]>1,] #5th is hgvspcolumn
print("Size of matrix of potential MNVs")
print(dim(allcodons.stats))

allcodons.stats$UID<-paste0(allcodons.stats[,1],allcodons.stats[,2], allcodons.stats[,3],allcodons.stats[,4],sep=".")
mafD$UID<-paste0(mafD[,symbolcol],mafD[,samplecol], mafD[,transcriptcol],mafD[,positioncol],sep=".")

allcodons.mnv<-merge(allcodons.stats,mafD,by="UID")
print("Done with putative MNV identification")

mnv<-getMNVs(unique(as.character(allcodons.mnv$UID)),allcodons.mnv)
mnv.df<-as.data.frame(do.call(rbind, mnv))
#Calculate frequencies, new MNV codon
mnv.df$MutReads1<-as.numeric(mnv.df$MutReads1)
mnv.df$MutReads2<-as.numeric(mnv.df$MutReads2)
mnv.df$RefReads1<-as.numeric(mnv.df$RefReads1)
mnv.df$RefReads2<-as.numeric(mnv.df$RefReads2)
mnv.df$Freq1<-mnv.df$MutReads1/(mnv.df$MutReads1+mnv.df$RefReads1)
mnv.df$Freq2<-mnv.df$MutReads2/(mnv.df$MutReads2+mnv.df$RefReads2)
mnv.df$MinAltCov<-with(mnv.df, pmin(mnv.df$MutReads1,mnv.df$MutReads2))
#Min 3 alt reads
mnv.df<-mnv.df[mnv.df$MinAltCov>2,]
pcor<-round(cor(mnv.df$Freq1,mnv.df$Freq2),2)
scor<-round(cor(mnv.df$Freq1,mnv.df$Freq2,method="spearman"),2)

sample.corp<-unlist(lapply(unique(mnv.df$Sample),function(x)cor(mnv.df[mnv.df$Sample==x,"Freq1"],mnv.df[mnv.df$Sample==x,"Freq2"])))

sample.cors<-unlist(lapply(unique(mnv.df$Sample),function(x)cor(mnv.df[mnv.df$Sample==x,"Freq1"],mnv.df[mnv.df$Sample==x,"Freq2"],method="spearman")))

sample.cor<-data.frame("Pearson"=sample.corp, "Spearman"=sample.cors)
sample.cor$Sample<-as.character(unlist(unique(mnv.df$Sample)))
sample.draw<-melt(sample.cor,id=c("Sample"))
colnames(sample.draw)<-c("Sample","Type","Correlation")

save(mnv.df,file="mnv.Rdat")


```


```{r MNV correction, echo=FALSE}

newcodons<-apply(mnv.df,1,combineCodons)
mnv.df$newCodon<-newcodons[1,]
mnv.df$newAA<-newcodons[2,]
consequences<-apply(mnv.df,1,determineConsequences)
mnv.df$MNVConsquence<-consequences
#Write
fname<-paste0(params$MAF,".MNV.txt")
#write.table(mnv.df,file=fname,sep="\t",quote=F,row.names = F)

```



###Plots describing the cohort

```{r correlationPlot, echo=FALSE}

#Overall correlation

ggplot(mnv.df,aes(x=Freq1,y=Freq2,col=log2(MinAltCov)))+geom_point()+ 
      ggtitle(paste("Same codon SNVs for all samples")) +
      annotate(geom="text", x=0.2, y=0.6, label=paste("Spearman",scor),color="red") +
      annotate(geom="text", x=0.2, y=0.63, label=paste("Pearson",pcor),color="red") +
      scale_x_continuous(expand = c(0, 0)) + scale_y_continuous(expand = c(0, 0))

ggplot(sample.draw,aes(x=Sample,y=Correlation, fill=Type))+geom_bar(position="dodge",stat="identity")+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

print("Summary of changes")
summary(as.factor(mnv.df$MNVConsquence))
sample.changes<-aggregate(UID ~ Sample+MNVConsquence,data=mnv.df,length)
colnames(sample.changes)<-c("Sample","MNVConsequence","Count")


#Correlations coefficients per sample

```

###Identify all probable MNVs: either very similar allele frequency or a difference of a single read for lower coverage mutations

[Corrected file `r fname`](`r fname`)

```{r identifyConfidence,echo=F}

#We are not certain these are not independent
mnv.df$diff<-abs(mnv.df$Freq1-mnv.df$Freq2)
mnv.df$diffF<-mnv.df$diff<0.05
mnv.df$diffR<-abs(mnv.df$MutReads1-mnv.df$MutReads2)
mnv.df$diffRF<-mnv.df$diffR<2
mnv.df$diff<-abs(mnv.df$Freq1-mnv.df$Freq2)
kable(sample.changes) 
ggplot(mnv.df,aes(x=Freq1,y=Freq2,col=(diffF|diffRF)))+geom_point()+ 
      ggtitle(paste("Misannotated MNVs at 5% AF difference or read diff<2")) +
      annotate(geom="text", x=0.2, y=0.6, label=paste("Spearman",scor),color="red") +
      annotate(geom="text", x=0.2, y=0.63, label=paste("Pearson",pcor),color="red") +
      scale_x_continuous(expand = c(0, 0)) + scale_y_continuous(expand = c(0, 0))

ggplot(mnv.df,aes(x=diff))+geom_histogram(bins=50) +
  ggtitle("Histogram of allele frequency difference of in codon SNVs")

#write.table(mnv.df,file=paste0(params$MAF,".mnvreport.txt"),sep="\t",row.names=F,quote=F)
write.table(mnv.df,file=fname,sep="\t",row.names=F,quote=F)

```
